drivers/note/notesnap_driver.c

```diff
@@ -0,0 +1,441 @@
+                              FAR volatile void *spinlock, int type);
+#endif
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SYSCALL
+static void notesnap_syscall_enter(FAR struct note_driver_s *drv, int nr,
+                                   int argc, va_list *ap);
+static void notesnap_syscall_leave(FAR struct note_driver_s *drv, int nr,
+                                   uintptr_t result);
+#endif
+#ifdef CONFIG_SCHED_INSTRUMENTATION_IRQHANDLER
+static void notesnap_irqhandler(FAR struct note_driver_s *drv, int irq,
+                                FAR void *handler, bool enter);
+#endif
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static const struct note_driver_ops_s g_notesnap_ops =
+{
+  NULL,
+  notesnap_start,
+  notesnap_stop,
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SWITCH
+  notesnap_suspend,
+  notesnap_resume,
+#  ifdef CONFIG_SMP
+  notesnap_cpu_start,
+  notesnap_cpu_started,
+  notesnap_cpu_pause,
+  notesnap_cpu_paused,
+  notesnap_cpu_resume,
+  notesnap_cpu_resumed,
+#  endif
+#endif
+#ifdef CONFIG_SCHED_INSTRUMENTATION_PREEMPTION
+  notesnap_premption,
+#endif
+#ifdef CONFIG_SCHED_INSTRUMENTATION_CSECTION
+  notesnap_csection,
+#endif
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SPINLOCKS
+  notesnap_spinlock,
+#endif
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SYSCALL
+  notesnap_syscall_enter,
+  notesnap_syscall_leave,
+#endif
+#ifdef CONFIG_SCHED_INSTRUMENTATION_IRQHANDLER
+  notesnap_irqhandler,
+#endif
+};
+
+static struct notesnap_s g_notesnap =
+{
+  {&g_notesnap_ops}
+};
+
+static const FAR char *g_notesnap_type[] =
+{
+  "NOTE_START",
+  "NOTE_STOP",
+  "NOTE_SUSPEND",
+  "NOTE_RESUME",
+  "NOTE_CPU_START",
+  "NOTE_CPU_STARTED",
+  "NOTE_CPU_PAUSE",
+  "NOTE_CPU_PAUSED",
+  "NOTE_CPU_RESUME",
+  "NOTE_CPU_RESUMED",
+  "NOTE_PREEMPT_LOCK",
+  "NOTE_PREEMPT_UNLOCK",
+  "NOTE_CSECTION_ENTER",
+  "NOTE_CSECTION_LEAVE",
+  "NOTE_SPINLOCK_LOCK",
+  "NOTE_SPINLOCK_LOCKED",
+  "NOTE_SPINLOCK_UNLOCK",
+  "NOTE_SPINLOCK_ABORT",
+  "NOTE_SYSCALL_ENTER",
+  "NOTE_SYSCALL_LEAVE",
+  "NOTE_IRQ_ENTER",
+  "NOTE_IRQ_LEAVE",
+  "NOTE_DUMP_STRING",
+  "NOTE_DUMP_BINARY"
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: notesnap_common
+ ****************************************************************************/
+
+static inline void notesnap_common(FAR struct tcb_s *tcb, uint8_t type,
+                                   uintptr_t args)
+{
+  if (g_notesnap.dumping)
+    {
+      return;
+    }
+
+  /* Atomic operation, equivalent to g_notesnap.index++; */
+
+  size_t index = atomic_fetch_add(&g_notesnap.index, 1);
+  FAR struct notesnap_chunk_s *note =
```

Comment:
C89 incompatible

Reply:
which part not C89 compatible?

Reply:
Variables should be declared at the beginning of the scope

Reply:
ok